src/Nethermind/Nethermind.Blockchain/OnChainTxWatcher.cs

```diff
@@ -58,22 +58,29 @@ private void OnBlockAddedToMain(object sender, BlockReplacementEventArgs e)
         private void ProcessBlock(Block block, Block? previousBlock)
         {
             _txPool.BlockGasLimit = block.GasLimit;
+            _txPool.CurrentBaseFee = block.Header.BaseFeePerGas;
             long transactionsInBlock = block.Transactions.Length;
             long discoveredForPendingTxs = 0;
             long discoveredForHashCache = 0;
             
             for (int i = 0; i < transactionsInBlock; i++)
             {
-                Keccak txHash = block.Transactions[i].Hash;
+                Transaction tx = block.Transactions[i];
+                Keccak txHash = tx.Hash;
+                
                 if (!_txPool.IsInHashCache(txHash))
                 {
                     discoveredForHashCache++;
                 }
-                if (!_txPool.RemoveTransaction(txHash, true))
+
+                if (!_txPool.RemoveTransaction(tx, true))
                 {
                     discoveredForPendingTxs++;
                 }
             }
+
+            _txPool.RemoveOrUpdateBuckets();
```

Comment:
Yeah, it was my initial idea to recalculate only buckets of addresses which changes in the last processed block.
But after every block we need to recalculate buckets with 1559 txs too. So I see three solutions.
1) in OnChainTxWatcher we can trigger recalculation only for buckets which changes. 1559 recalculation can be triggered by TxPool.CurrentBaseFee change. We have then every bucket recalculated exactly when there is a need for it, but we have two triggers for recalculations in two different places and some buckets can be unnecessarily recalculated two times. And as we use lockers, there can be some threats because of too many concurrent actions.
2) in OnChainTxWatcher we can trigger recalculating changed buckets and buckets with 1559 txs. It was my previous implementation, before changes. But then we need to do a snapshot of all txpool and check all txs if it is 1559 or not, so we are doing plenty of work which can be avoided.
3) in OnChainTxWatcher we can trigger recalculating of all buckets - like in current shape of this branch. Then in TxPool we are recalculating every single bucket, even those without 1559 and without changes. But after London most of txs is expected to be 1559 so we will probably need to recalculate almost every bucket anyway. And thanks to this solution we are up-to-date with potential changes of account balances - if some address received incoming tx and has higher balance than before, can potentially has higher GasBottleneck.
I believe the third solution is best because of simplification, avoiding potential double-recalculation and having all txs as updated as possible

About triggering recalculation - TxPool listening to BlockAddedToMain can cause recalulation of buckets before removing processd transactions in OnChainTxWatcher. Same story with listening to TxPool.CurrentBaseFee. We want to recalculate buckets in it's shape after removing processed transactions, so triggering it from inside of OnChainTxWatcher looks optimal

And about renaming - RemoveOrUpdateBuckets is removing bucket if balance is insufficient or otherwise updating it. We can change it for something more abstract, but I don't have idea about potential new name. Maybe just "UpdateBuckets", and possibility of removal will be hidden in implementation?